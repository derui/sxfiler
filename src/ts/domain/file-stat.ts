import * as bigInt from "big-integer";

// define type and class of file stat
export interface Capability {
  readonly writable: boolean;
  readonly readable: boolean;
  readonly executable: boolean;
}

export interface Mode {
  readonly owner: Capability;
  readonly group: Capability;
  readonly others: Capability;
}

// information object of file
export class FileStat {
  constructor(
    public readonly mode: Mode,
    public readonly uid: number,
    public readonly gid: number,
    public readonly atime: Date,
    public readonly ctime: Date,
    public readonly mtime: Date,
    public readonly size: string,
    public readonly isDirectory: boolean,
    public readonly isFile: boolean,
    public readonly isSymlink: boolean
  ) {}

  get sizeAsBigInt(): bigInt.BigInteger {
    return bigInt(this.size);
  }
}
