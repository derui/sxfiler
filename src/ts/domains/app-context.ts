import { UIContext } from "@/types/ui-context";

export type AppContext = {
  readonly current: UIContext;
  readonly subContexts: UIContext[];
};

/**
   create AppContext
 */
export function createAppContext({
  current,
  subContexts,
}: {
  current: UIContext;
  subContexts?: UIContext[];
}): AppContext {
  return {
    current,
    subContexts: Array.from(subContexts || []),
  };
}

/**
   change context
 */
export function changeCurrent(context: UIContext) {
  return (state: AppContext) =>
    createAppContext({
      current: context,
      subContexts: state.subContexts,
    });
}

/**
   Add a context to AppContext
 */
export function addSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.add(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}

/**
   remove the context from AppContext
 */
export function removeSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.delete(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}
