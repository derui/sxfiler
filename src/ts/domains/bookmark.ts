/**
   Type of bookmark
 */
export type Bookmark = {
  readonly id: string;
  readonly path: string;
  readonly order: number;
};

type FactoryArg = Pick<Bookmark, keyof Bookmark>;

/**
 * Create bookmark from factory argument
 */
export const createBookmark = function createBookmark({ id, path, order }: FactoryArg): Bookmark {
  return {
    id,
    path,
    order,
  };
};
