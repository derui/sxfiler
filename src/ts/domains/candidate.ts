export type CandidateObject = {
  readonly id: string;
  readonly value: string;
  readonly start: number;
  readonly length: number;
};

export type Candidate = CandidateObject & {
  /**
   * Get splitted candidate as 3 pieces by matched string that
   * has the one of before start of match, matched, and after matched.
   */
  splitByInput(): [string, string, string];

  /**
   * Get plain object.
   */
  plain(): CandidateObject;
};

type FactoryArg = Pick<CandidateObject, "id" | "value"> & Partial<Pick<CandidateObject, "start" | "length">>;

/**
 * Create candidate from candidate object.
 */
export const createCandidate = ({ id, value, start = 0, length = 0 }: FactoryArg): Candidate => {
  return {
    id,
    value,
    start: Math.max(0, Math.min(value.length - 1, start)),
    length: Math.max(0, Math.min(value.length, length)),
    splitByInput() {
      const beforeMatched = value.substring(0, start);
      const matched = value.substring(start, start + length);
      const afterMatched = value.substring(start + length);

      return [beforeMatched, matched, afterMatched];
    },

    plain() {
      return { id: this.id, value: this.value, start: this.start, length: this.length };
    },
  };
};
