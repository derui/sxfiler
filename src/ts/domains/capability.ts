// Capability with method
export type Capability = CapabilityObject & {
  /** Return new capability that allows to write */
  allowToWrite(): Capability;

  /** Return new capability that allows to read */
  allowToRead(): Capability;

  /** Return new capability that allows to execute */
  allowToExecute(): Capability;

  /** Return new capability that disallows to write */
  disallowToWrite(): Capability;

  /** Return new capability that disallows to read */
  disallowToRead(): Capability;

  /** Return new capability that disallows to execute */
  disallowToExecute(): Capability;

  /** Return plain CapabilityObject */
  plain(): CapabilityObject;
};

// Capability of a target
export type CapabilityObject = {
  readonly writable: boolean;
  readonly readable: boolean;
  readonly executable: boolean;
};

type CreateCapabilityArg = {
  writable: boolean;
  readable: boolean;
  executable: boolean;
};

function plain(this: Capability): CapabilityObject {
  return { writable: this.writable, readable: this.readable, executable: this.executable };
}

/**
   Create capability from argument
 */
export const createCapability = ({ writable, readable, executable }: CreateCapabilityArg): Capability => {
  return {
    writable,
    readable,
    executable,

    plain,

    /** Return new capability that allows to write */
    allowToWrite() {
      return { ...this, writable: true };
    },

    /** Return new capability that allows to read */
    allowToRead() {
      return { ...this, readable: true };
    },

    /** Return new capability that allows to execute */
    allowToExecute() {
      return { ...this, executable: true };
    },

    /** Return new capability that disallows to write */
    disallowToWrite() {
      return { ...this, writable: false };
    },

    /** Return new capability that disallows to read */
    disallowToRead() {
      return { ...this, readable: false };
    },

    /** Return new capability that disallows to execute */
    disallowToExecute() {
      return { ...this, executable: false };
    },
  };
};

/**
   Return new capability that have not any work.
 */
export const emptyCapability = (): Capability =>
  createCapability({ writable: false, readable: false, executable: false });

/**
   Return new capability that have all capability on all roles.
 */
export const fullCapability = (): Capability => createCapability({ writable: true, readable: true, executable: true });
