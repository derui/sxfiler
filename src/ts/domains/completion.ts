import { Candidate, CandidateObject, createCandidate } from "./candidate";
import { between } from "@/utils";

export type CompletionObject = {
  // current key map
  readonly cursor: number;
  readonly candidates: CandidateObject[];
};

export type Completion = {
  // current key map
  readonly cursor: number;
  readonly candidates: Candidate[];

  /**
     replace candidates with new candidates
   */
  replaceCandidates(candidates: Candidate[]): Completion;

  /**
     move cursor to amount value
   */
  moveCursor(amount: number): Completion;

  /**
     return plain object
   */
  plain(): CompletionObject;
};

function replaceCandidates(this: Completion, candidates: CandidateObject[]): Completion {
  return { ...this, candidates: candidates.map(createCandidate) };
}

function moveCursor(this: Completion, amount: number): Completion {
  return { ...this, cursor: between(this.cursor + amount, this.candidates.length - 1) };
}

function plain(this: Completion): CompletionObject {
  return { cursor: this.cursor, candidates: this.candidates.map(v => v.plain()) };
}

/** return empty state */
export const createCompletion = ({ cursor, candidates }: Partial<CompletionObject>): Completion => {
  if (cursor && candidates) {
    cursor = between(cursor, candidates.length);
  } else if (candidates) {
    cursor = 0;
  } else {
    cursor = 0;
    candidates = [];
  }
  return { cursor, candidates: candidates.map(createCandidate), replaceCandidates, plain, moveCursor };
};
