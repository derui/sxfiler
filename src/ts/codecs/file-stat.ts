import { FileStat as Domain, createFileStat } from "@/domains/file-stat";
import { createCapability, emptyCapability } from "@/domains/capability";
import { createMode } from "@/domains/mode";
import { FileStat, Mode, Capability } from "../generated/filer_pb";

const defaultMode = createMode({
  owner: emptyCapability(),
  group: emptyCapability(),
  others: emptyCapability(),
});

const encodeCap = function encodeCapability(cap: Capability) {
  return createCapability({
    writable: cap.writable,
    readable: cap.readable,
    executable: cap.executable,
  });
};

const encodeMode = function encodeMode(mode: Mode) {
  const owner = mode.owner;
  const group = mode.group;
  const others = mode.others;

  return createMode({
    owner: owner ? encodeCap(Capability.create(owner)) : emptyCapability(),
    group: group ? encodeCap(Capability.create(group)) : emptyCapability(),
    others: others ? encodeCap(Capability.create(others)) : emptyCapability(),
  });
};

/**
   encode node object from RPC to frontend domain.

   @param obj JSON representation for node
   @return Node object
 */
export const encode = function encode(obj: FileStat): Domain {
  const mode = obj.mode;
  return createFileStat({
    mode: mode ? encodeMode(Mode.create(mode)) : defaultMode,
    uid: obj.uid,
    gid: obj.gid,
    atime: obj.atime,
    ctime: obj.ctime,
    mtime: obj.mtime,
    size: obj.size,
    isDirectory: obj.isDirectory,
    isFile: obj.isFile,
    isSymlink: obj.isSymlink,
  });
};
