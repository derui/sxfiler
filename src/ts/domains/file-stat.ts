import bigInt from "big-integer";
import { Mode, createMode } from "./mode";
import { createCapability } from "./capability";

export type FileStat = {
  readonly mode: Mode;
  readonly uid: number;
  readonly gid: number;
  readonly atime: Date;
  readonly ctime: Date;
  readonly mtime: Date;
  readonly size: string;
  readonly isDirectory: boolean;
  readonly isFile: boolean;
  readonly isSymlink: boolean;
};

export type FactoryArg = {
  mode: Mode;
  uid: number;
  gid: number;
  atime: string;
  ctime: string;
  mtime: string;
  size: string;
  isDirectory: boolean;
  isFile: boolean;
  isSymlink: boolean;
};

export const sizeAsBigInt = function sizeAsBigInt(state: FileStat): bigInt.BigInteger {
  return bigInt(state.size);
};

/**
 * create FileStat
 * @param arg arguments of factory
 */
export const createFileStat = function createFileStat(arg: FactoryArg): FileStat {
  const { atime, ctime, mtime, mode, ...rest } = arg;
  return {
    ...rest,
    mode: createMode({
      group: createCapability(mode.group),
      owner: createCapability(mode.owner),
      others: createCapability(mode.others),
    }),
    atime: new Date(Number(atime)),
    ctime: new Date(Number(ctime)),
    mtime: new Date(Number(mtime)),
  };
};
