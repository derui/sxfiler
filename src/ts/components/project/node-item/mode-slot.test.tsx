import * as React from "react";
import renderer from "react-test-renderer";
import T from "./mode-slot";

import { Capability, Mode } from "../../../domains/file-stat";

describe("Project", () => {
  describe("Node Item", () => {
    describe("Mode slot", () => {
      it("should print name simply", () => {
        const mode = new Mode();
        const tree = renderer.create(<T mode={mode} isDirectory={false} isSymlink={false} />).toJSON();

        expect(tree).toMatchSnapshot();
      });

      it("should print directory mode", () => {
        const mode = new Mode({ owner: new Capability({ readable: true }) });
        const tree = renderer.create(<T mode={mode} isDirectory={true} isSymlink={false} />).toJSON();

        expect(tree).toMatchSnapshot();
      });

      it("should print symlink mode", () => {
        const mode = new Mode({ owner: new Capability({ readable: true }) });
        const tree = renderer.create(<T mode={mode} isDirectory={false} isSymlink={true} />).toJSON();

        expect(tree).toMatchSnapshot();
      });

      it("should print symlink when directory and symlink are true", () => {
        const mode = new Mode({ owner: new Capability({ readable: true }) });
        const tree = renderer.create(<T mode={mode} isDirectory={true} isSymlink={true} />).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });
});
