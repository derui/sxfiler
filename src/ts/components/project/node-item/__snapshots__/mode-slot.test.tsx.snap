// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Node Item Mode slot should print directory mode 1`] = `
<span
  className="mode"
>
  dr--------
</span>
`;

exports[`Project Node Item Mode slot should print name simply 1`] = `
<span
  className="mode"
>
  ----------
</span>
`;

exports[`Project Node Item Mode slot should print symlink mode 1`] = `
<span
  className="mode"
>
  lr--------
</span>
`;

exports[`Project Node Item Mode slot should print symlink when directory and symlink are true 1`] = `
<span
  className="mode"
>
  lr--------
</span>
`;
